File: glean-core/csharp/GleanTests/Metrics/CounterMetricTypeTest.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using Mozilla.Glean.Testing;
using System;
using System.IO;
using Xunit;
using static Mozilla.Glean.Glean;

namespace Mozilla.Glean.Tests.Metrics
{
    public class CounterMetricTypeTest
    {
        public CounterMetricTypeTest()
        {
            // Get a random test directory just for this single test.
            string tempDataDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());

            // In xUnit, the constructor will be called before each test. This
            // feels like a natural place to initialize / reset Glean.
            GleanInstance.Reset(
	            applicationId: "org.mozilla.csharp.tests",
                applicationVersion: "1.0-test",
                uploadEnabled: true,
                configuration: new Configuration(),
                dataDir: tempDataDir
                );
        }

        [Fact]
        public void APISavesToStorage()
        {
            Private.CounterMetricType counterMetric = new Private.CounterMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "counter_metric",
                sendInPings: new string[] { "store1" }
            );

            Assert.False(counterMetric.TestHasValue());

            // Add to the counter a couple of times with a little delay.  The first call will check
            // calling add() without parameters to test increment by 1.
            counterMetric.Add();

            // Check that the count was incremented and properly recorded.
            Assert.True(counterMetric.TestHasValue());
            Assert.Equal(1, counterMetric.TestGetValue());

            counterMetric.Add(10);
            // Check that count was incremented and properly recorded.  This second call will check
            // calling add() with 10 to test increment by other amount
            Assert.True(counterMetric.TestHasValue());
            Assert.Equal(11, counterMetric.TestGetValue());
        }

        [Fact]
        public void DisabledCountersMustNotRecordData()
        {
            Private.CounterMetricType counterMetric = new Private.CounterMetricType(
                category: "telemetry",
                disabled: true,
                lifetime: Private.Lifetime.Application,
                name: "counter_metric",
                sendInPings: new string[] { "store1" }
            );

            // Attempt to store the counter.
            counterMetric.Add();
            // Check that nothing was recorded.
            Assert.False(counterMetric.TestHasValue(), "Counters must not be recorded if they are disabled");
        }

        [Fact]
        public void TestGetValueThrows()
        {
            Private.CounterMetricType counterMetric = new Private.CounterMetricType(
                category: "telemetry",
                disabled: true,
                lifetime: Private.Lifetime.Application,
                name: "counter_metric",
                sendInPings: new string[] { "store1" }
            );
            Assert.Throws<NullReferenceException>(() => counterMetric.TestGetValue());
        }

        [Fact]
        public void APISavesToSecondaryPings()
        {
            Private.CounterMetricType counterMetric = new Private.CounterMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "counter_metric",
                sendInPings: new string[] { "store1", "store2" }
            );

            // Add to the counter a couple of times with a little delay.  The first call will check
            // calling add() without parameters to test increment by 1.
            counterMetric.Add();

            // Check that the count was incremented and properly recorded for the second ping.
            Assert.True(counterMetric.TestHasValue("store2"));
            Assert.Equal(1, counterMetric.TestGetValue("store2"));

            counterMetric.Add(10);
            // Check that count was incremented and properly recorded for the second ping.
            // This second call will check calling add() with 10 to test increment by other amount
            Assert.True(counterMetric.TestHasValue("store2"));
            Assert.Equal(11, counterMetric.TestGetValue("store2"));
        }

        [Fact]
        public void NegativeValuesAreNotCounted()
        {
            Private.CounterMetricType counterMetric = new Private.CounterMetricType(
                category: "telemetry",
                disabled: false,
                lifetime: Private.Lifetime.Application,
                name: "counter_metric",
                sendInPings: new string[] { "store1" }
            );

            // Increment to 1 (initial value)
            counterMetric.Add();

            // Check that the count was incremented
            Assert.True(counterMetric.TestHasValue("store1"));
            Assert.Equal(1, counterMetric.TestGetValue("store1"));

            counterMetric.Add(-10);
            // Check that count was NOT incremented.
            Assert.True(counterMetric.TestHasValue("store1"));
            Assert.Equal(1, counterMetric.TestGetValue("store1"));
            Assert.Equal(1, counterMetric.TestGetNumRecordedErrors(ErrorType.InvalidValue));
        }
}
}

File: glean-core/android/src/test/java/mozilla/telemetry/glean/private/CounterMetricTypeTest.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package mozilla.telemetry.glean.private

/* This file is based on the tests in the Glean android-components implementation.
 *
 * Care should be taken to not reorder elements in this file so it will be easier
 * to track changes in Glean android-components.
 */

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import mozilla.telemetry.glean.testing.ErrorType
import mozilla.telemetry.glean.testing.GleanTestRule
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class CounterMetricTypeTest {

    @get:Rule
    val gleanRule = GleanTestRule(ApplicationProvider.getApplicationContext())

    @Test
    fun `The API saves to its storage engine`() {
        // Define a 'counterMetric' counter metric, which will be stored in "store1"
        val counterMetric = CounterMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "counter_metric",
                sendInPings = listOf("store1"),
            ),
        )

        assertNull(counterMetric.testGetValue())

        // Add to the counter a couple of times with a little delay.  The first call will check
        // calling add() without parameters to test increment by 1.
        counterMetric.add()

        // Check that the count was incremented and properly recorded.
        assertEquals(1, counterMetric.testGetValue())

        counterMetric.add(10)
        // Check that count was incremented and properly recorded.  This second call will check
        // calling add() with 10 to test increment by other amount
        assertEquals(11, counterMetric.testGetValue())
    }

    @Test
    fun `disabled counters must not record data`() {
        // Define a 'counterMetric' counter metric, which will be stored in "store1".  It's disabled
        // so it should not record anything.
        val counterMetric = CounterMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "counter_metric",
                sendInPings = listOf("store1"),
            ),
        )

        // Attempt to store the counter.
        counterMetric.add()
        // Check that nothing was recorded.
        assertNull(
            "Counters must not be recorded if they are disabled",
            counterMetric.testGetValue(),
        )
    }

    @Test
    fun `testGetValue() returns null if nothing is stored`() {
        val counterMetric = CounterMetricType(
            CommonMetricData(
                disabled = true,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "counter_metric",
                sendInPings = listOf("store1"),
            ),
        )
        assertNull(counterMetric.testGetValue())
    }

    @Test
    fun `The API saves to secondary pings`() {
        // Define a 'counterMetric' counter metric, which will be stored in "store1" and "store2"
        val counterMetric = CounterMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "counter_metric",
                sendInPings = listOf("store1", "store2"),
            ),
        )

        // Add to the counter a couple of times with a little delay.  The first call will check
        // calling add() without parameters to test increment by 1.
        counterMetric.add()

        // Check that the count was incremented and properly recorded for the second ping.
        assertEquals(1, counterMetric.testGetValue("store2"))

        counterMetric.add(10)
        // Check that count was incremented and properly recorded for the second ping.
        // This second call will check calling add() with 10 to test increment by other amount
        assertEquals(11, counterMetric.testGetValue("store2"))
    }

    @Test
    fun `negative values are not counted`() {
        // Define a 'counterMetric' counter metric, which will be stored in "store1"
        val counterMetric = CounterMetricType(
            CommonMetricData(
                disabled = false,
                category = "telemetry",
                lifetime = Lifetime.APPLICATION,
                name = "counter_metric",
                sendInPings = listOf("store1"),
            ),
        )

        // Increment to 1 (initial value)
        counterMetric.add()

        // Check that the count was incremented
        assertEquals(1, counterMetric.testGetValue("store1"))

        counterMetric.add(-10)
        // Check that count was NOT incremented.
        assertEquals(1, counterMetric.testGetValue("store1"))
        assertEquals(1, counterMetric.testGetNumRecordedErrors(ErrorType.INVALID_VALUE))
    }
}
